### ISSUE TITLE
Unmarshal() fails with an error "duplicated struct field name"

### ISSUE THREAD

**Issue by ikedas** [goccy/go-yaml#352]
## Version

v1.10.0

## Description

When there is a field with a name that matches the custom key of a tag attached to another field in the structure, Unmarshal() fails with an error "duplicated struct field name".

構造体の別のフィールドに付けられたタグのカスタムキーと一致する名前のフィールドがある場合に「重複したフィールド名」としてUnmarshal()が失敗する。

### Test code
``` go
package main

import (
        "github.com/goccy/go-yaml"
        "log"
)

type Config struct {
        Domain string
        Host   string `yaml:"domain"`
}

func main() {
        config := Config{}
        buf := "---\ndomain:\n"
        if err := yaml.Unmarshal([]byte(buf), &config); err != nil {
                log.Fatal(err)
        }
}
```
### Result
``` code
2023/03/06 17:23:25 duplicated struct field name domain:
    github.com/goccy/go-yaml.structFieldMap
        /home/user/go/pkg/mod/github.com/goccy/go-yaml@v1.10.0/struct.go:124
```

**Comment by shuheiktgw** [comment]
Thank you for reporting the issue! Unfortunately, this isn't a bug but rather an intentional design choice. In Go, if you omit the YAML field name, it implicitly assumes it, which has resulted in the Config structs having duplicated field names. I've verified that gopkg.in/yaml.v3 behaves the same way.

**Comment by Kyslik** [comment]
Please see https://github.com/go-yaml/yaml/issues/467. 

TLDR: `encoding/json` allows "duplicated" struct fields.

**Comment by goccy** [comment]
@Kyslik  Yes, of course I know that encoding/json behaves that way. I’m the author of [go-json](https://github.com/goccy/go-json), and go-json is fully compatible with encoding/json. So, I know the specification well, and I implemented all the features myself.

However, applying this behavior to go-yaml requires careful consideration. The encoding/json specification is not exactly intuitive, so if we allow duplicate fields, I believe there needs to explicitly declaration to specify which field should be used. Whether this is done via tags or options needs to be carefully evaluated.
